### PR TITLE
Integrate Rockset as a vector store

### PR DESCRIPTION
**Pull request by kwadhwa18**
Integrates [Rockset](https://rockset.com/docs/) as a vector store.

**Description**

Rockset is a real-time analytics database which enables queries on massive, semi-structured data without operational burden. Rockset is serverless and fully managed. It offloads the work of managing configuration, cluster provisioning, denormalization, and shard / index management. Rockset is also SOC 2 Type II compliant and offers encryption at rest and in flight, securing and protecting any sensitive data. Most teams can ingest data into Rockset and start executing queries in less than 15 minutes.

This PR is similar to https://github.com/langchain-ai/langchain/pull/6216.

   * Added integration tests
   * Added docs & example program
   * Linted & formatted

### CODE REVIEW

File: langchain/src/vectorstores/rockset.ts [hunk 1]
```diff
@@ -0,0 +1,418 @@
+import { MainApi } from '@rockset/client';
+import type { CreateCollectionRequest } from "@rockset/client/dist/codegen/api.d.ts";
```
Comment (jacoblee93) [inline langchain/src/vectorstores/rockset.ts:2]: I think you will need to add this path here:

https://github.com/langchain-ai/langchainjs/blob/main/langchain/scripts/check-tree-shaking.js#L24

File: langchain/src/vectorstores/rockset.ts [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+    this.name = this.constructor.name;
+  }
+}
+
+/**
+ * Functions to measure vector distance/similarity by.
+ * See https://rockset.com/docs/vector-functions/#vector-distance-functions
+ * @enum SimilarityMetric
+ */
+export const enum SimilarityMetric {
```
Comment (jacoblee93) [inline langchain/src/vectorstores/rockset.ts:40]: We can't use `enums` for tree shaking reasons - can you change to a type?

File: langchain/src/vectorstores/rockset.ts [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+    }
+    const similarityKey = "similarity"
+    const _filter = filter ?? this.filter;
+    return (await this.client.queries.query({
+      sql: {
+        query: `
+          SELECT
+            * EXCEPT("${this.embeddingKey}"),
+            "${this.textKey}",
+            ${this.similarityMetric}([${query}], "${this.embeddingKey}") AS "${similarityKey}"
```
Comment (jacoblee93) [inline langchain/src/vectorstores/rockset.ts:236]: Looks like a SQL injection?

File: examples/src/indexes/vector_stores/rockset.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import * as rockset from "@rockset/client";
+import { ChatOpenAI } from "langchain/chat_models/openai";
+import { RetrievalQAChain } from "langchain/chains";
```
Comment (jacoblee93) [inline examples/src/indexes/vector_stores/rockset.ts:3]: You're welcome to use this chain but we've moved towards expression language:

https://js.langchain.com/docs/expression_language/cookbook/retrieval

File: examples/src/indexes/vector_stores/rockset.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import { OpenAIEmbeddings } from "langchain/embeddings/openai";
+import { RocksetStore } from "langchain/vectorstores/rockset";
+import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
+import { readFileSync } from "fs";
+
+export const run = async () => {
+    const store = await RocksetStore.withNewCollection(
+        new OpenAIEmbeddings(),
+        {
+            client: rockset.default.default(
```
Comment (jacoblee93) [inline examples/src/indexes/vector_stores/rockset.ts:13]: Run `yarn format` at top level.

File: examples/src/indexes/vector_stores/rockset.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    );
+
+    const model = new ChatOpenAI({ modelName: "gpt-3.5-turbo" });
+    const chain = RetrievalQAChain.fromLLM(model, store.asRetriever());
+    const text = readFileSync("state_of_the_union.txt", "utf8");
+    const docs = await (new RecursiveCharacterTextSplitter()).createDocuments([text]);
+
+    await store.addDocuments(docs);
+    const response = await chain.call({
+        query: "What is America's role in Ukraine?"
```
Comment (jacoblee93) [inline examples/src/indexes/vector_stores/rockset.ts:28]: Will this return anything? I think the state of the union in examples is old.

**Comment by jacoblee93** [comment]
Thanks for this - see comments

File: langchain/src/vectorstores/rockset.ts (hunk 3 above)
Comment (kwadhwa18) [reply langchain/src/vectorstores/rockset.ts]: parameterized query. filter string is tricky, it is something like `age>10` etc - rockset doesn't allow such string params in the API request

File: examples/src/indexes/vector_stores/rockset.ts (hunk 5 above)
Comment (kwadhwa18) [reply examples/src/indexes/vector_stores/rockset.ts:13]: done

**Comment by kwadhwa18** [comment]
cc @jacoblee93 can you please take a look again? I addressed most

File: langchain/src/vectorstores/rockset.ts (hunk 3 above)
Comment (jacoblee93) [reply langchain/src/vectorstores/rockset.ts]: Ah that's super unfortunate. Oh well.

**Comment by jacoblee93** [comment]
Thank you!

**Review by jacoblee93 — APPROVED** [review]

**Pull request merged**